semantics/src/main/kotlin/com/strumenta/kolasu/semantics/linking/ReferenceResolver.kt

```diff
@@ -0,0 +1,138 @@
+package com.strumenta.kolasu.semantics.linking
+
+import com.strumenta.kolasu.model.PossiblyNamed
+import com.strumenta.kolasu.model.ReferenceByName
+import com.strumenta.kolasu.semantics.indexing.InMemoryIndex
+import com.strumenta.kolasu.semantics.indexing.Index
+import com.strumenta.kolasu.semantics.indexing.IndexReader
+import com.strumenta.kolasu.utils.sortBySubclassesFirst
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty1
+import kotlin.reflect.safeCast
+
+/**
+ * Type-safe builder to create resolver instances.
+ * @param index optional index for external symbols
+ * @param cache optional storage to cache computations
```

Comment:
It's not really optional though

Reply:
You're right, optional is probably not the best word. I will change to specify that if no index is provided, an initially empty in-memory one will be used. What do you think?

Reply:
Commented

Reply:
Modified documentation with:
```kotlin
@param index index for external symbols (default: [InMemoryIndex])
@param cache storage to cache computations (default: [InMemoryReferenceResolverCache])
```